File: src/pointer.cpp
#include "pointer.hpp"
#include "iostream"

void pointer_increment(int *i, int j, int *k) {
    *i += 3;
    *k =  *i + j;
}

void modify_array(unsigned int * p_arr, unsigned int p_arr_len)
{
    std::cout << "Provided array value: ";
    for(unsigned int i = 0; i < p_arr_len; i++)
    {
        std::cout << *(p_arr + i) << "  "; 
    }
    std::cout << std::endl;

    std::cout << "Now modifying inside function" << std::endl;
    for(unsigned int i = 0; i < p_arr_len; i++)
    {
        *(p_arr + i) = *(p_arr + i) * 2; 
    }
}

int Color_Input(int * color_ip,
                int ** color_op,
                Color_Info * color_info)
{
    std::cout << "Color Number: " << color_info->color_number << std::endl;
    std::cout << "Inside function color_ip" << std::endl;
    for (unsigned int i = 0; i< color_info->color_number; i++)
        std::cout << *(color_ip + i) << " ";
    std::cout << std::endl;
    int* local_op_ = new int[9];

	for (unsigned int i = 0; i< color_info->color_number; i++)
	{
		local_op_[i] = *(color_ip + i);
	}
    std::cout << "Inside function local_op_" << std::endl;
    for (unsigned int i = 0; i< color_info->color_number; i++)
        std::cout << *(local_op_ + i) << " ";
    std::cout << std::endl;

	*color_op = &local_op_[0];
    std::cout << "Inside function color_info->color_box_number: " << *(color_info->color_box_number) << std::endl;
    *(color_info->color_box_number) = 8;
    return 1;
}

void Check_Struct(Child * obj)
{
    obj->age = 12;
    *(obj->number) = 123;
}

File: example/py/pointer.py
# Py file to demonstrate the usage of binded pointer.cpp

print("#"*10, "pointer","#"*10)

import numpy as np
import pybind_cpp_module

# Following function call pointer_increment()
p_1 = 12
p_2 = 23
p_3 = 9
p_1, p_2, p_3 = pybind_cpp_module.pointer_increment(p_1, p_2, p_3)
print("p_1: ",p_1, " p_2: ",p_2, " p_3: ",p_3)

# Following function call modify_array()
# Create a NumPy array
x_list = [1,2,5,6,4,6]
x_array = np.array(x_list, dtype=np.uint32)

# Get the length of the array
arr_len = len(x_list)

# Call the C++ function modify_array
pybind_cpp_module.modify_array(x_array)
print("Now array: ", x_array)

# Call the C++ function Color_Input
# It requires different arg to call than C++ function as py-binding 
# has written differently

color_info = pybind_cpp_module.Color_Info()
color_info.color_company_name = "Blue"
color_info.color_number = 9
color_info.color_box_number = 1
color_info.color_code = 2
color_ip = []
color_op = []
ret = 0
for i in range (9):
    color_ip.append(2 * i)

ret, color_op = pybind_cpp_module.Color_Input(color_ip, color_info)

print("ret: ", ret)
print("color_op: ", color_op)

# Call the C++ function Check_Struct
child_obj = pybind_cpp_module.Child()
child_obj.number = 8
ret_obj = pybind_cpp_module.Check_Struct(child_obj)
print("Age in py: ", ret_obj.age)
print("NUmber in py: ", ret_obj.number)
